# Exploratory Data Analysis for Data Science


#### This project demonstrates the application of Data Analysis using Python from fundamental to advance. First we will take a look at the following: 



#### First Part

- Arithmetic Operators 
- Comparison Operators 
- Logic Operators
- Data Types and Functions 

#### Second Part
- Load and Read Data 
- Using Numpy array and Pandas Library
- Data Visualisation 
- Data Cleaning and Comouting of Missig Data 

## Application of Arithmetic Operators 

In [8]:
1

1

In [9]:
1.0

1.0

In [10]:
1+1

2

In [11]:
1*3

3

In [12]:
1/2.0

0.5

In [13]:
2**4

16

In [14]:
2 + 3 * 5 + 5

22

In [15]:
(2+3) * (5+5)

50

In [16]:
4%2

0

In [17]:
5%2

1

In [18]:
8 % 2

0

In [19]:
var = 2

In [20]:
var

2

In [21]:
x = 2
y = 3

In [22]:
x + y

5

In [23]:
x = x+x

In [24]:
x

4

In [25]:
x = x+x

In [26]:
x

8

In [27]:
name_of_var = 12

## Data Types Examples and Functions

In [29]:
'single quote'

'single quote'

In [30]:
"this is a string"

'this is a string'

In [31]:
"I can't go"

"I can't go"

In [32]:
x = 'hello'

In [33]:
x

'hello'

In [34]:
print(x)

hello


In [35]:
num = 12
name = 'Sam'

In [36]:
print('My number is {} and my name is {}'.format(num,name))

My number is 12 and my name is Sam


In [37]:
num = 12
name = 'Sam'

In [38]:
print('My number is {one} and my name is {two} more{one}'.format(one=num,two=name))

My number is 12 and my name is Sam more12


In [39]:
s = 'hello'

In [40]:
s[4]

'o'

In [41]:
s = 'abcdefghijk'

In [42]:
s[0:]

'abcdefghijk'

In [43]:
s[:3]

'abc'

In [44]:
s[3:6]

'def'

### List a sequence of element 

In [46]:
[1,2,3]

[1, 2, 3]

In [47]:
['a','b','c']

['a', 'b', 'c']

In [48]:
my_list = ['a','b','c']

In [49]:
my_list.append('d')

In [50]:
my_list

['a', 'b', 'c', 'd']

In [51]:
my_list[0]

'a'

In [52]:
my_list[1:3]

['b', 'c']

In [53]:
my_list[0]='New'

In [54]:
my_list

['New', 'b', 'c', 'd']

In [55]:
nest = [1,2,[3,4]]

In [56]:
nest

[1, 2, [3, 4]]

In [57]:
nest[2][1]

4

In [58]:
nest = [1,2,3,[4,5,['target']]]

In [59]:
print(nest[3][2][0])

target


### Dictionary 

In [61]:
d = {'key1':"value", 'key2':123}

In [62]:
d['key1']

'value'

In [63]:
d['key2']

123

In [64]:
d = {'k1':[1,2,3]}

In [65]:
my_list = d['k1']

In [66]:
my_list[0]

1

In [67]:
d['k1'][0]

1

In [68]:
d = {'k1':{'innerkey':[1,2,3]}}

In [69]:
d['k1']['innerkey'][1]

2

### Booleans Data Types

In [71]:
True 

True

In [72]:
False

False

In [73]:
my_list = [1,2,3]

In [74]:
my_list[0]

1

In [75]:
t = (1,2,3)

In [76]:
t[0]

1

In [77]:
my_list [0] = 'New' 

In [78]:
my_list

['New', 2, 3]

In [79]:
t 

(1, 2, 3)

- **Sets is defined by only unique numbers**

In [81]:
{1,2,3}

{1, 2, 3}

In [82]:
{1,1,1,2,2,2,3,3,3}

{1, 2, 3}

In [83]:
set([1,1,1,2,2,2,5,5,5,5,6,6,6])

{1, 2, 5, 6}

In [84]:
s = {1,2,3}

In [85]:
s.add(5)

In [86]:
s

{1, 2, 3, 5}

### Comparison Operators

In [88]:
1>2

False

In [89]:
1<2

True

In [90]:
1 >=2

False

In [91]:
1<=1

True

In [92]:
1==1

True

In [93]:
1 == 2

False

In [94]:
1 != 3

True

In [95]:
'hi' == 'bye'

False

In [96]:
'hi'!= 'bye'

True

### Logic Operators 

In [98]:
(1 < 2) and (2 > 3)

False

In [99]:
(1 < 2) or (2 > 3) or (1 == 1)

True

In [100]:
True and True

True

In [101]:
True and False

False

In [102]:
True or False 

True

- **Code block**

In [104]:
if 1 < 2: 
    print('yep!')

yep!


In [105]:
if True :
    print('perform code')

perform code


In [106]:
if True: 
    x = 2+2

In [107]:
x

4

### Multiple Conditions 

In [109]:
if 1 == 2:
    print('First')
else:
    print('last')

last


In [110]:
if 1 != 2:
    print('First')
else:
    print('last')

First


In [111]:
if 1 == 2:
    print('First')
elif 4==4:
    print('second')
elif 3 == 3:
    print('Middle')
else:
    print('last')


second


### For and Why Loops 

In [113]:
seq = [1,2,3,4,5]

In [114]:
for item in seq:
    print(item)

1
2
3
4
5


In [115]:
seq = [1,2,3,4,5]

In [116]:
for num in seq:
    print('hello')

hello
hello
hello
hello
hello


### For infinit i: While Loop 

In [118]:
i = 1
while i < 5:
    print('i is: {}'.format(i))
    i = i+1

i is: 1
i is: 2
i is: 3
i is: 4


In [119]:
seq = [1,2,3,4,5]

In [120]:
for x in seq:
    print(x)

1
2
3
4
5


### Range

In [122]:
range(0,5)

range(0, 5)

In [123]:
for x in range(0,5):
    print(x)

0
1
2
3
4


In [124]:
list(range(0,5))

[0, 1, 2, 3, 4]

In [125]:
list(range(10))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

### List Comprehension

In [127]:
x = [1,2,3,4]

In [128]:
out = []

for num in x: 
    out.append(num**2)

In [129]:
print(out)

[1, 4, 9, 16]


### Transforming a for list code into a list comperehension 

In [131]:
[num**2 for num in x]

[1, 4, 9, 16]

In [132]:
out = [num**2 for num in x]

In [133]:
out

[1, 4, 9, 16]

### Functions uses key word: 'def'

In [135]:
def my_func(param1):
    print(param1)

In [136]:
my_func('helo')

helo


In [137]:
def my_func(name):
    print('Hello ' +name)

In [138]:
my_func('Jose')

Hello Jose


In [139]:
def my_func(name= 'Default Name'):
    print('Hello ' + name)

In [140]:
my_func()

Hello Default Name


In [141]:
my_func(name='Jose')

Hello Jose


In [142]:
my_func

<function __main__.my_func(name='Default Name')>

In [143]:
def square(num):
    """
    THIS IS A DOCSTRING.
    CAN GO MULTIPLE LINES.
    THIS FUNCTION SQUARES A NUMBER.
    """
    return num**2

In [144]:
output = square(2)

In [145]:
output

4

- **NOTE: You can use 'shift + tab' to get suggestions of the functions**

In [148]:
def times2(var):
    return var*2

In [149]:
times2(5)

10

- **map()**

In [151]:
seq = [1,2,3,4,5]

In [152]:
map(times2, seq)

In [153]:
list(map(times2,seq))

[2, 4, 6, 8, 10]

### Lambda Expression 

In [155]:
t = lambda var:var*2

In [156]:
t(6)

12

- **using lambda for map expression**

- Using the previous defined map 

In [159]:
seq = [1,2,3,4,5]

In [160]:
list(map(lambda num: num*3, seq))

[3, 6, 9, 12, 15]

### Filter Function: filter out even numbers 

In [162]:
list(filter(lambda num: num%2 ==0, seq))

[2, 4]

### Methods 

In [164]:
s = 'hello my is Sam'

In [165]:
s.lower()

'hello my is sam'

In [166]:
s.upper()

'HELLO MY IS SAM'

### Split Function 

In [168]:
s.split()

['hello', 'my', 'is', 'Sam']

In [169]:
tweet = 'Go Sports! #Sports'

In [170]:
tweet.split()

['Go', 'Sports!', '#Sports']

In [171]:
tweet.split('#')

['Go Sports! ', 'Sports']

In [172]:
tweet.split('#')[1]

'Sports'

In [173]:
d = {'k1': 1, 'k2': 2}

In [174]:
d

{'k1': 1, 'k2': 2}

In [175]:
d.keys()

dict_keys(['k1', 'k2'])

In [176]:
d.items()

dict_items([('k1', 1), ('k2', 2)])

In [177]:
d.values()

dict_values([1, 2])

In [178]:
lst = [1,2,3]

In [179]:
lst.pop()

3

In [180]:
lst

[1, 2]

In [181]:
lst

[1, 2]

In [182]:
lst = [1,2,3,4,5]

In [183]:
item = lst.pop()

In [184]:
lst

[1, 2, 3, 4]

In [185]:
item

5

In [186]:
first = lst.pop(0)

In [187]:
lst

[2, 3, 4]

In [188]:
first

1

In [189]:
lst.append('new')

In [190]:
lst

[2, 3, 4, 'new']

### To check for an item inside a list 

In [192]:
'x' in [1,2,3]

False

In [193]:
'x' in ['x','y','z']

True

### Using tuple unpacking 

In [195]:
x = [(1,2),(3,4),(5,6)]

In [196]:
x

[(1, 2), (3, 4), (5, 6)]

In [197]:
x[0]

(1, 2)

In [198]:
x[0][0]

1

In [199]:
x[0][1]

2

In [200]:
for item in x:
    print(item)

(1, 2)
(3, 4)
(5, 6)


In [201]:
for (a,b) in x:
    print(b)

2
4
6


In [202]:
for a,b in x:
    print(a)
    print(b)

1
2
3
4
5
6


## Install Python Libraries for Data Analysis 

**Using NumPy Arrays**

- Numpy
- Pandas
- Matplotlib
Python Data Analysis Libray: Numpy Arrays 

In [204]:
my_list = [1,2,3]

In [205]:
import numpy as np

In [206]:
arr = np.array(my_list)

In [207]:
arr

array([1, 2, 3])

In [208]:
my_math = [[1,2,3],[4,5,6],[7,8,9]]

In [209]:
np.array(my_math)

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [210]:
np.arange(0,11)

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [211]:
np.arange(0,11,2)

array([ 0,  2,  4,  6,  8, 10])

In [212]:
np.zeros(3)

array([0., 0., 0.])

In [213]:
np.zeros((2,3))

array([[0., 0., 0.],
       [0., 0., 0.]])

In [214]:
np.ones(4)

array([1., 1., 1., 1.])

In [215]:
np.ones((3,4))

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

- **Using `linspace`: returns evenely distributed numbers**

In [217]:
np.linspace(0,5,100)

array([0.        , 0.05050505, 0.1010101 , 0.15151515, 0.2020202 ,
       0.25252525, 0.3030303 , 0.35353535, 0.4040404 , 0.45454545,
       0.50505051, 0.55555556, 0.60606061, 0.65656566, 0.70707071,
       0.75757576, 0.80808081, 0.85858586, 0.90909091, 0.95959596,
       1.01010101, 1.06060606, 1.11111111, 1.16161616, 1.21212121,
       1.26262626, 1.31313131, 1.36363636, 1.41414141, 1.46464646,
       1.51515152, 1.56565657, 1.61616162, 1.66666667, 1.71717172,
       1.76767677, 1.81818182, 1.86868687, 1.91919192, 1.96969697,
       2.02020202, 2.07070707, 2.12121212, 2.17171717, 2.22222222,
       2.27272727, 2.32323232, 2.37373737, 2.42424242, 2.47474747,
       2.52525253, 2.57575758, 2.62626263, 2.67676768, 2.72727273,
       2.77777778, 2.82828283, 2.87878788, 2.92929293, 2.97979798,
       3.03030303, 3.08080808, 3.13131313, 3.18181818, 3.23232323,
       3.28282828, 3.33333333, 3.38383838, 3.43434343, 3.48484848,
       3.53535354, 3.58585859, 3.63636364, 3.68686869, 3.73737

- **Note: one dimensions will have one bracket, 2 dimensions will 2 brackets and 3 dimensions 3 brackets**

- linspace will take the 3rd arguments as the number of points you want while arange will take that arguement as the step sides you want.
- Identify matrix using numpy: useful matrix when dealing with linear algebra (two dimensional square matrix)


In [221]:
np.eye(4)

array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])

In [222]:
np.random.rand(5)

array([0.53777588, 0.64848962, 0.13590154, 0.07391076, 0.85251126])

In [223]:
np.random.rand(5,5)

array([[0.07692304, 0.62892999, 0.98185521, 0.28241614, 0.5737995 ],
       [0.90761261, 0.25214766, 0.84117383, 0.64780804, 0.11994458],
       [0.90104807, 0.02414782, 0.96317702, 0.84872455, 0.36267033],
       [0.21441413, 0.36679034, 0.62043342, 0.96665749, 0.82573595],
       [0.79671223, 0.39763867, 0.23739767, 0.8072046 , 0.55433964]])

- **Returning a sample from a randomly distributed number**

In [225]:
np.random.randn(2)

array([-1.38571378, -0.09882887])

In [226]:
np.random.randn(4,4)

array([[-0.74059879,  0.85384502,  0.9708423 , -0.92872186],
       [ 0.87794375, -2.60755385,  0.75305875,  0.03238095],
       [ 0.81491567,  0.14771192,  0.3592814 , -0.05453487],
       [ 1.19296095, -0.24220127,  0.4772561 , -1.05322438]])

- **To return a ramdom number from a selection of randomly distributed numbers, for example selecting a number in a population of 100 distributions.**

In [228]:
np.random.randint(1,100)

49

- **For example selcting 10 random integers from a 100 distributed samples.** 


In [230]:
np.random.randint(1,100,10)

array([40, 38, 46, 89, 73, 74, 99, 61, 40, 98])

In [231]:
arr = np.arange (25)

In [232]:
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24])

In [233]:
ranarr = np.random.randint(0,50,10)

In [234]:
ranarr

array([15, 23, 48, 28, 29, 10, 34, 20, 40, 22])

- **Using Reshape to reshape the sample to the number of dimensions need.** Note the sample will need to be able to form a complete matrix.

In [236]:
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24])

In [237]:
arr.reshape(5,5)

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24]])

- **Calling maximum and minimum values from a number of randonmly distributed array of numbers.**

In [239]:
ranarr.max()

48

In [240]:
ranarr.min()

10

In [241]:
ranarr

array([15, 23, 48, 28, 29, 10, 34, 20, 40, 22])

In [242]:
ranarr.argmax()

2

- This implies that the max vlaue 48 is located at an index number 2

In [244]:
ranarr.argmin()

5

- This also implies that the minmum value (1) is located at an index number 5 

- **Using attributes to figure out the shape of a vector** 

In [247]:
arr.shape

(25,)

- one dimension 

In [249]:
arr = arr.reshape(5,5)

In [250]:
arr

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24]])

- Reshaped to 2 dimensions

- Check for Data type

In [253]:
arr.dtype

dtype('int64')

- This shows a 64 type intergers

In [255]:
from numpy.random import randint

In [256]:
randint(2,10)

5

### Numpy Indexing and Selection

In [258]:
import numpy as np


In [259]:
arr = np.arange (0,11)

In [260]:
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [261]:
arr[8]

8

In [262]:
arr[1:5]

array([1, 2, 3, 4])

In [263]:
arr[0:5]

array([0, 1, 2, 3, 4])

In [264]:
arr[:6]

array([0, 1, 2, 3, 4, 5])

In [265]:
arr[5:]

array([ 5,  6,  7,  8,  9, 10])

### Broadcasting

In [267]:
arr[0:5] = 100

In [268]:
arr

array([100, 100, 100, 100, 100,   5,   6,   7,   8,   9,  10])

In [269]:
arr = np.arange(0,11)

In [270]:
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [271]:
slice_of_arr = arr[0:6]

In [272]:
slice_of_arr

array([0, 1, 2, 3, 4, 5])

In [273]:
slice_of_arr[:] = 99

In [274]:
slice_of_arr

array([99, 99, 99, 99, 99, 99])

In [275]:
arr

array([99, 99, 99, 99, 99, 99,  6,  7,  8,  9, 10])

- **Tips: To make sure that the slice does not affecct the original array use the function copy.** 

In [277]:
arr_copy = arr.copy()

In [278]:
arr

array([99, 99, 99, 99, 99, 99,  6,  7,  8,  9, 10])

In [279]:
arr_copy

array([99, 99, 99, 99, 99, 99,  6,  7,  8,  9, 10])

In [280]:
arr_copy[:] = 100

In [281]:
arr_copy

array([100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100])

In [282]:
arr

array([99, 99, 99, 99, 99, 99,  6,  7,  8,  9, 10])

- **Use copy function to change the array without affecring the original array values.**
- **For 2D array known as Matrics: 2D arrays must have 2 block brackets.** 

In [285]:
import numpy as np 

In [286]:
arr_2d = np.array([[5,10,15],[20,25,30],[35,40,45]])

In [287]:
arr_2d

array([[ 5, 10, 15],
       [20, 25, 30],
       [35, 40, 45]])

- Applying the double bracket

In [289]:
arr_2d

array([[ 5, 10, 15],
       [20, 25, 30],
       [35, 40, 45]])

In [290]:
arr_2d[0][0]

5

In [291]:
arr_2d[0]

array([ 5, 10, 15])

In [292]:
arr_2d[1][1]

25

In [293]:
arr_2d[2][1]

40

- **Using the comma single bracket notation.** 

In [295]:
arr_2d[2,1]

40

In [296]:
arr_2d[1,2]

30

In [297]:
arr_2d[1][2]

30

- **Use colon for slice notation to grab certain section of the entire 2D elements.** 

In [299]:
arr_2d[:2,1:]

array([[10, 15],
       [25, 30]])

In [300]:
arr_2d[:2]

array([[ 5, 10, 15],
       [20, 25, 30]])

In [301]:
arr_2d[1:]

array([[20, 25, 30],
       [35, 40, 45]])

### Conditional Selection: Very Important 

In [303]:
arr = np.arange(1,11)

In [304]:
arr

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [305]:
arr > 5

array([False, False, False, False, False,  True,  True,  True,  True,
        True])

In [306]:
arr.dtype

dtype('int64')

In [307]:
bool_arr = arr > 5

In [308]:
bool_arr

array([False, False, False, False, False,  True,  True,  True,  True,
        True])

In [309]:
arr[bool_arr]

array([ 6,  7,  8,  9, 10])

- Notice that this conditional select all the elements that happens to be true. In this case greater than 5.

In [311]:
arr[arr>5]

array([ 6,  7,  8,  9, 10])

In [312]:
arr[arr<3]

array([1, 2])

In [313]:
arr_2d = np.arange(50).reshape(5,10)

In [314]:
arr_2d

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14, 15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24, 25, 26, 27, 28, 29],
       [30, 31, 32, 33, 34, 35, 36, 37, 38, 39],
       [40, 41, 42, 43, 44, 45, 46, 47, 48, 49]])

In [315]:
arr_2d[:2,1]

array([ 1, 11])

In [316]:
arr_2d[1:,:3]

array([[10, 11, 12],
       [20, 21, 22],
       [30, 31, 32],
       [40, 41, 42]])

In [317]:
arr_2d[1:3,:4]

array([[10, 11, 12, 13],
       [20, 21, 22, 23]])

In [318]:
arr_2d[1:3,]

array([[10, 11, 12, 13, 14, 15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24, 25, 26, 27, 28, 29]])

In [319]:
arr_2d[1:3,3:5]

array([[13, 14],
       [23, 24]])

In [320]:
arr_2d[3:,7:9]

array([[37, 38],
       [47, 48]])

### Numpy Operations: 
- Array with array
- Array with scalars
- Universal array functions 

In [322]:
import numpy as np

In [323]:
arr = np.arange(0,11)

In [324]:
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [325]:
arr + arr

array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18, 20])

- Notice that the numbers have doubled; and this can also be applied with subtraction and multiplications 

In [327]:
arr - arr

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [328]:
arr * arr

array([  0,   1,   4,   9,  16,  25,  36,  49,  64,  81, 100])

In [329]:
arr + 100

array([100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110])

In [330]:
arr * 100

array([   0,  100,  200,  300,  400,  500,  600,  700,  800,  900, 1000])

In [331]:
arr - 100

array([-100,  -99,  -98,  -97,  -96,  -95,  -94,  -93,  -92,  -91,  -90])

In [332]:
arr / arr

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: RuntimeWarning: invalid value encountered in true_divide
  """Entry point for launching an IPython kernel.


array([nan,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.])

- **This is to keep in mind when working with numbers dividing by zero.** 

In [334]:
arr ** 2

array([  0,   1,   4,   9,  16,  25,  36,  49,  64,  81, 100])

## Universal Array Functions: 
- Which are simply mathematical operations that can be usedin array operations to broadcast across the entire array

In [336]:
np.sqrt(arr)

array([0.        , 1.        , 1.41421356, 1.73205081, 2.        ,
       2.23606798, 2.44948974, 2.64575131, 2.82842712, 3.        ,
       3.16227766])

- **This takes the square root of the array.**

In [338]:
np.exp(arr)

array([1.00000000e+00, 2.71828183e+00, 7.38905610e+00, 2.00855369e+01,
       5.45981500e+01, 1.48413159e+02, 4.03428793e+02, 1.09663316e+03,
       2.98095799e+03, 8.10308393e+03, 2.20264658e+04])

In [339]:
np.max(arr)

10

In [340]:
arr.max()

10

- **Metrix Function.**

In [342]:
np.sin(arr)

array([ 0.        ,  0.84147098,  0.90929743,  0.14112001, -0.7568025 ,
       -0.95892427, -0.2794155 ,  0.6569866 ,  0.98935825,  0.41211849,
       -0.54402111])

In [343]:
np.log(arr)

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: RuntimeWarning: divide by zero encountered in log
  """Entry point for launching an IPython kernel.


array([      -inf, 0.        , 0.69314718, 1.09861229, 1.38629436,
       1.60943791, 1.79175947, 1.94591015, 2.07944154, 2.19722458,
       2.30258509])

In [345]:
import numpy as np

In [346]:
arr_2d = np.arange(50).reshape(10,5)

In [347]:
arr_2d

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24],
       [25, 26, 27, 28, 29],
       [30, 31, 32, 33, 34],
       [35, 36, 37, 38, 39],
       [40, 41, 42, 43, 44],
       [45, 46, 47, 48, 49]])

In [348]:
arr_2d[2,2]

12

In [349]:
arr_2d[:3,1:]

array([[ 1,  2,  3,  4],
       [ 6,  7,  8,  9],
       [11, 12, 13, 14]])

In [350]:
arr_2d[4:6,2:4]

array([[22, 23],
       [27, 28]])

- **Creating a matrix of 10 by 10 that goes from 0.1 to 1.0.** 

In [352]:
np.arange(1,101).reshape(10,10)/100

array([[0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.1 ],
       [0.11, 0.12, 0.13, 0.14, 0.15, 0.16, 0.17, 0.18, 0.19, 0.2 ],
       [0.21, 0.22, 0.23, 0.24, 0.25, 0.26, 0.27, 0.28, 0.29, 0.3 ],
       [0.31, 0.32, 0.33, 0.34, 0.35, 0.36, 0.37, 0.38, 0.39, 0.4 ],
       [0.41, 0.42, 0.43, 0.44, 0.45, 0.46, 0.47, 0.48, 0.49, 0.5 ],
       [0.51, 0.52, 0.53, 0.54, 0.55, 0.56, 0.57, 0.58, 0.59, 0.6 ],
       [0.61, 0.62, 0.63, 0.64, 0.65, 0.66, 0.67, 0.68, 0.69, 0.7 ],
       [0.71, 0.72, 0.73, 0.74, 0.75, 0.76, 0.77, 0.78, 0.79, 0.8 ],
       [0.81, 0.82, 0.83, 0.84, 0.85, 0.86, 0.87, 0.88, 0.89, 0.9 ],
       [0.91, 0.92, 0.93, 0.94, 0.95, 0.96, 0.97, 0.98, 0.99, 1.  ]])

- **A differnt approach to create the same metrix.**

In [354]:
np.linspace(0.01,1,100).reshape(10,10)

array([[0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.1 ],
       [0.11, 0.12, 0.13, 0.14, 0.15, 0.16, 0.17, 0.18, 0.19, 0.2 ],
       [0.21, 0.22, 0.23, 0.24, 0.25, 0.26, 0.27, 0.28, 0.29, 0.3 ],
       [0.31, 0.32, 0.33, 0.34, 0.35, 0.36, 0.37, 0.38, 0.39, 0.4 ],
       [0.41, 0.42, 0.43, 0.44, 0.45, 0.46, 0.47, 0.48, 0.49, 0.5 ],
       [0.51, 0.52, 0.53, 0.54, 0.55, 0.56, 0.57, 0.58, 0.59, 0.6 ],
       [0.61, 0.62, 0.63, 0.64, 0.65, 0.66, 0.67, 0.68, 0.69, 0.7 ],
       [0.71, 0.72, 0.73, 0.74, 0.75, 0.76, 0.77, 0.78, 0.79, 0.8 ],
       [0.81, 0.82, 0.83, 0.84, 0.85, 0.86, 0.87, 0.88, 0.89, 0.9 ],
       [0.91, 0.92, 0.93, 0.94, 0.95, 0.96, 0.97, 0.98, 0.99, 1.  ]])

In [355]:
mat = np.arange(1,26).reshape(5,5)

In [356]:
mat

array([[ 1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10],
       [11, 12, 13, 14, 15],
       [16, 17, 18, 19, 20],
       [21, 22, 23, 24, 25]])

In [357]:
mat[2:,1:]

array([[12, 13, 14, 15],
       [17, 18, 19, 20],
       [22, 23, 24, 25]])

In [358]:
mat[:3,1]

array([ 2,  7, 12])

In [359]:
mat[:3,1:2]

array([[ 2],
       [ 7],
       [12]])

- **Notice the difference in the above slice matrices, one is one dimensional and the other is 2 (two) dimensional.** 

In [361]:
mat[4,:]

array([21, 22, 23, 24, 25])

In [362]:
mat[-1,:]

array([21, 22, 23, 24, 25])

In [363]:
mat[3:,:]

array([[16, 17, 18, 19, 20],
       [21, 22, 23, 24, 25]])

In [364]:
mat[3:5,:]

array([[16, 17, 18, 19, 20],
       [21, 22, 23, 24, 25]])

- **Get the sum of all the values in the matrices.**

In [366]:
np.sum(mat)

325

In [367]:
mat.sum()

325

- **Get the standard deviation of all the values in the matrices.**

In [369]:
np.std(mat)

7.211102550927978

In [370]:
mat.std()

7.211102550927978

- **Get the sum of all the column in the matrices.**

In [372]:
mat.sum(axis=0)

array([55, 60, 65, 70, 75])

## Pandas for Data Analysis

- Use either "conda install pandas" or "pip install pandas" to install pandas package library


In [376]:
import numpy as np

In [377]:
import pandas as pd

In [378]:
labels = ['a', 'b', 'c']
my_data = [10,20,30]
arr = np.array(my_data)
d = {'a':10, 'b':20, 'c':30}

In [379]:
pd.Series(data = my_data)

0    10
1    20
2    30
dtype: int64

In [380]:
pd.Series(data=my_data,index=labels)

a    10
b    20
c    30
dtype: int64

- **Notice that the data lable is now index a, b, c.**

In [382]:
pd.Series(my_data,labels)

a    10
b    20
c    30
dtype: int64

- **Other method to create a Series.**

In [384]:
pd.Series(arr)

0    10
1    20
2    30
dtype: int64

In [385]:
pd.Series(arr,labels)

a    10
b    20
c    30
dtype: int64

In [386]:
pd.Series(d)

a    10
b    20
c    30
dtype: int64

In [387]:
d

{'a': 10, 'b': 20, 'c': 30}

- **Note series can hold a variety of data type: like strings for example which makes it diff from numpy array.**

In [389]:
pd.Series(data=labels)

0    a
1    b
2    c
dtype: object

In [390]:
pd.Series(data=[sum,print,len])

0      <built-in function sum>
1    <built-in function print>
2      <built-in function len>
dtype: object

- **The key to using a series is understanding its index.**

### How to grab an informatuion from a series: In this task my index is country names

In [393]:
ser1 = pd.Series([1,2,3,4],['USA','Germany','USSR','Japan'])

In [406]:
ser1

USA        1
Germany    2
USSR       3
Japan      4
dtype: int64

In [407]:
ser2 = pd.Series([1,2,5,4],['USA','Germany','Italy','Japan'])

In [408]:
ser2

USA        1
Germany    2
Italy      5
Japan      4
dtype: int64

- **To grab information from the series: Keep in mind the differnt data type between strings and integers.**

In [410]:
ser1['USA']

1

In [411]:
ser3 = pd.Series(data=labels)

In [412]:
ser3

0    a
1    b
2    c
dtype: object

In [413]:
ser3[0]

'a'

In [414]:
ser1 + ser2

Germany    4.0
Italy      NaN
Japan      8.0
USA        2.0
USSR       NaN
dtype: float64

- **Notice that where it didnt find a country match it will put an Nan. Also the intergers are converted to floats so that you don't have loose any object.**

### DataFrame

In [417]:
import numpy as np
import pandas as pd

In [418]:
from numpy.random import randn

In [419]:
np.random.seed(101)

In [420]:
df = pd.DataFrame(randn(5,4),['A','B','C','D','E'],['W','X','Y','Z'])

In [421]:
df

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

- **Grabbing a data from the table.**

In [423]:
df['W']

A    2.706850
B    0.651118
C   -2.018168
D    0.188695
E    0.190794
Name: W, dtype: float64

In [424]:
type(df['W'])

pandas.core.series.Series

In [425]:
type(df)

pandas.core.frame.DataFrame

In [426]:
df.W

A    2.706850
B    0.651118
C   -2.018168
D    0.188695
E    0.190794
Name: W, dtype: float64

In [427]:
df['X']

A    0.628133
B   -0.319318
C    0.740122
D   -0.758872
E    1.978757
Name: X, dtype: float64

In [428]:
df[['W','Z']]

W         Z
A  2.706850  0.503826
B  0.651118  0.605965
C -2.018168 -0.589001
D  0.188695  0.955057
E  0.190794  0.683509

- **Creating a new column.** 

In [430]:
df['new'] = df['W'] + df['Y']

In [431]:
df

W         X         Y         Z       new
A  2.706850  0.628133  0.907969  0.503826  3.614819
B  0.651118 -0.319318 -0.848077  0.605965 -0.196959
C -2.018168  0.740122  0.528813 -0.589001 -1.489355
D  0.188695 -0.758872 -0.933237  0.955057 -0.744542
E  0.190794  1.978757  2.605967  0.683509  2.796762

- **To remove column: Use 'df.drop', use shift+tab to get suggestions.** 

In [433]:
df.drop('new', axis=1)

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

In [434]:
df

W         X         Y         Z       new
A  2.706850  0.628133  0.907969  0.503826  3.614819
B  0.651118 -0.319318 -0.848077  0.605965 -0.196959
C -2.018168  0.740122  0.528813 -0.589001 -1.489355
D  0.188695 -0.758872 -0.933237  0.955057 -0.744542
E  0.190794  1.978757  2.605967  0.683509  2.796762

- **Notice that if you call df, you will still have the new column. This means that you have to specify an argument.**
- The 'arguement inplace' allows pandas to retain information incase you accidentally remove information. 

In [436]:
df.drop('new', axis=1, inplace=True)

In [437]:
df

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

- **To drop a row: 'E' for example.** 

In [439]:
df.drop('E', axis=0) 

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057

In [440]:
df

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

- **Note to completely delete the row 'E' you have to put inplace=True command.**

- To check for shape

In [443]:
df.shape

(5, 4)

In [444]:
df

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

### Selecting Rows:
- There are two methods to do this
- The first method is the 'loc' method known as the 'location method'

In [446]:
df.loc['A']

W    2.706850
X    0.628133
Y    0.907969
Z    0.503826
Name: A, dtype: float64

### The Second Method:
- The second method is the 'iloc' method which is used to grab the index location 

In [448]:
df.iloc[0]

W    2.706850
X    0.628133
Y    0.907969
Z    0.503826
Name: A, dtype: float64

In [449]:
df.iloc[3]

W    0.188695
X   -0.758872
Y   -0.933237
Z    0.955057
Name: D, dtype: float64

- **Selecting subsect of rows.**

In [451]:
df.loc['B','Y']

-0.8480769834036315

In [452]:
df.loc[['A','B'],['W','Y']]

W         Y
A  2.706850  0.907969
B  0.651118 -0.848077

- **This approach allows you to select the subsect of A & B rows with their corresponding cloumn W & Y in the the DataFrame.**
- Note in the DataFrame, to drop a column, specify 'axis=1' and 'axis=0' for rows.

### Using Conditional Selection for DataFrames

In [456]:
import numpy as np
import pandas as pd

In [457]:
from numpy.random import randn

In [458]:
np.random.seed(101)

In [459]:
df=pd.DataFrame(randn(5,4),['A','B','C','D','E'],['W','X','Y','Z'])

In [460]:
df

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

- **Conditional selection using bracket notation.**

In [462]:
df > 0

W      X      Y      Z
A   True   True   True   True
B   True  False  False   True
C  False   True   True  False
D   True  False  False   True
E   True   True   True   True

In [463]:
booldf =df>0

In [464]:
df[booldf]

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118       NaN       NaN  0.605965
C       NaN  0.740122  0.528813       NaN
D  0.188695       NaN       NaN  0.955057
E  0.190794  1.978757  2.605967  0.683509

In [465]:
df[df>0]

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118       NaN       NaN  0.605965
C       NaN  0.740122  0.528813       NaN
D  0.188695       NaN       NaN  0.955057
E  0.190794  1.978757  2.605967  0.683509

- **To return only the values where the subsets are true.** 

In [467]:
df

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

In [468]:
df['W'] > 0 

A     True
B     True
C    False
D     True
E     True
Name: W, dtype: bool

In [469]:
df['W']

A    2.706850
B    0.651118
C   -2.018168
D    0.188695
E    0.190794
Name: W, dtype: float64

- Now to filter out rows that are true based on the above information

In [471]:
df [df['W']>0]

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

In [472]:
df [df['W']>0]

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

In [473]:
df[df['Z']<0]

W         X         Y         Z
C -2.018168  0.740122  0.528813 -0.589001

In [474]:
resultdf= df[df['W']>0]

In [475]:
resultdf['X']

A    0.628133
B   -0.319318
D   -0.758872
E    1.978757
Name: X, dtype: float64

- To perform the operations all in one step 

In [477]:
df[df['W']>0]['X']

A    0.628133
B   -0.319318
D   -0.758872
E    1.978757
Name: X, dtype: float64

In [478]:
df[df['W']>0][['Y','X']]

Y         X
A  0.907969  0.628133
B -0.848077 -0.319318
D -0.933237 -0.758872
E  2.605967  1.978757

In [479]:
boolser = df['W'] >0
result = df[boolser]
result

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

In [484]:
mycols = ['Y','X']

In [485]:
result[mycols]

Y         X
A  0.907969  0.628133
B -0.848077 -0.319318
D -0.933237 -0.758872
E  2.605967  1.978757

- **Dealing with multiple conditions such as W>0 and Y>1 for example.**
- Note 'and' operator can only take one condition at a time, for example we can see the below   

In [487]:
df[(df['W']>0) and (df['Y']>1)]

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

- This is because 'and' operator can only take one condition at a time for example 

In [489]:
True and False 

False

In [490]:
True and True 

True

- **So pandas allows us to use multiple conditions using '&'

In [492]:
df[(df['W']>0) & (df['Y']>1)]

W         X         Y         Z
E  0.190794  1.978757  2.605967  0.683509

- Anothor operation is the use of 'slash operator' which is 'shift + tab' abobve enter key. 
- Note: using 'or' will return and error

In [494]:
df[(df['W']>0) | (df['Y']>1)]

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

- index: to reset the data to the original data 

In [496]:
df.reset_index()

index         W         X         Y         Z
0     A  2.706850  0.628133  0.907969  0.503826
1     B  0.651118 -0.319318 -0.848077  0.605965
2     C -2.018168  0.740122  0.528813 -0.589001
3     D  0.188695 -0.758872 -0.933237  0.955057
4     E  0.190794  1.978757  2.605967  0.683509

In [497]:
df

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

- **Creating a new index
- Note '.split()' is a function that can be used for separating blank spaces

In [499]:
newind = 'CA NY WY OR CO'.split()

In [500]:
newind

['CA', 'NY', 'WY', 'OR', 'CO']

- Now adding these to my column in the DataFrame

In [513]:
df['States'] = newind

In [514]:
df

W         X         Y         Z States
A  2.706850  0.628133  0.907969  0.503826     CA
B  0.651118 -0.319318 -0.848077  0.605965     NY
C -2.018168  0.740122  0.528813 -0.589001     WY
D  0.188695 -0.758872 -0.933237  0.955057     OR
E  0.190794  1.978757  2.605967  0.683509     CO

In [516]:
df.drop('States', axis=1)

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

In [517]:
import numpy as np
import pandas as pd

In [518]:
from numpy.random import randn

In [519]:
np.random.seed(101)

In [520]:
df = pd.DataFrame(randn(5,4),['A','B','C','D','E'],['W','X','Y','Z'])

In [521]:
df

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

- Add a new column 

In [523]:
newind = 'CA NY WY OR CO'.split()

In [524]:
newind

['CA', 'NY', 'WY', 'OR', 'CO']

- Now add these to my column the DataFrame named 'States'

In [526]:
df['States']=newind

In [527]:
df

W         X         Y         Z States
A  2.706850  0.628133  0.907969  0.503826     CA
B  0.651118 -0.319318 -0.848077  0.605965     NY
C -2.018168  0.740122  0.528813 -0.589001     WY
D  0.188695 -0.758872 -0.933237  0.955057     OR
E  0.190794  1.978757  2.605967  0.683509     CO

- Set the 'States' as index

In [529]:
df.set_index('States')

W         X         Y         Z
States                                        
CA      2.706850  0.628133  0.907969  0.503826
NY      0.651118 -0.319318 -0.848077  0.605965
WY     -2.018168  0.740122  0.528813 -0.589001
OR      0.188695 -0.758872 -0.933237  0.955057
CO      0.190794  1.978757  2.605967  0.683509

- Notice that no setting the inplace argument will return to original data when called

In [531]:
df

W         X         Y         Z States
A  2.706850  0.628133  0.907969  0.503826     CA
B  0.651118 -0.319318 -0.848077  0.605965     NY
C -2.018168  0.740122  0.528813 -0.589001     WY
D  0.188695 -0.758872 -0.933237  0.955057     OR
E  0.190794  1.978757  2.605967  0.683509     CO

- To retain the 'States index' as unchanged, type 'inplace=True' after the States argunent

### Working with Multi-Index and Higher DataFrame

In [534]:
import numpy as np 
import pandas as pd

In [535]:
# Index Levels
outside = ['G1','G1','G1','G2','G2','G2']
inside = [1,2,3,1,2,3]
hier_index = list(zip(outside,inside))
hier_index = pd.MultiIndex.from_tuples(hier_index)

In [536]:
outside

['G1', 'G1', 'G1', 'G2', 'G2', 'G2']

In [537]:
inside

[1, 2, 3, 1, 2, 3]

In [538]:
list(zip(outside,inside))

[('G1', 1), ('G1', 2), ('G1', 3), ('G2', 1), ('G2', 2), ('G2', 3)]

In [539]:
hier_index

MultiIndex([('G1', 1),
            ('G1', 2),
            ('G1', 3),
            ('G2', 1),
            ('G2', 2),
            ('G2', 3)],
           )

- Create a DataFrame with 6 rows and 2 columns, 
- Set index to the hier_index and columns to a list A and B

In [541]:
df = pd.DataFrame(randn(6,2),hier_index,['A','B'])

In [542]:
df

A         B
G1 1  0.302665  1.693723
   2 -1.706086 -1.159119
   3 -0.134841  0.390528
G2 1  0.166905  0.184502
   2  0.807706  0.072960
   3  0.638787  0.329646

- **Notice that I now have 2 levels of DataFrame G1 and G2.**

In [544]:
df.loc['G1']

A         B
1  0.302665  1.693723
2 -1.706086 -1.159119
3 -0.134841  0.390528

- Now grab the first 2 data from G1.
- The trick here is to call from outside to inside

In [546]:
df.loc['G1'].loc[1]

A    0.302665
B    1.693723
Name: 1, dtype: float64

- Name the index for example the outside index 'Group' and inside index 'Num'

In [548]:
df.index.names = ['Groups', 'Num']

In [549]:
df

A         B
Groups Num                    
G1     1    0.302665  1.693723
       2   -1.706086 -1.159119
       3   -0.134841  0.390528
G2     1    0.166905  0.184502
       2    0.807706  0.072960
       3    0.638787  0.329646

- Filter the information from the Data 

In [551]:
df.loc['G2'].loc[2]['B']

0.07295967531703869

In [552]:
df.loc['G1'].loc[3]['A']

-0.13484072060601238

- Cross-sections of rows and columns 

In [554]:
df.xs('G1')

A         B
Num                    
1    0.302665  1.693723
2   -1.706086 -1.159119
3   -0.134841  0.390528

- Select row 1 in G1 and  row 1 in G2

In [556]:
df.xs(1, level='Num')

A         B
Groups                    
G1      0.302665  1.693723
G2      0.166905  0.184502

### Dealing with Missing Data using Pandas

- Using Pandas to fill in missing Data
- Key A has a list 1,2 and use 'np.nan' to identify missing values

In [559]:
import numpy as np
import pandas as pd

In [560]:
d = {'A':[1,2,np.nan],'B':[5,np.nan,np.nan],'C':[1,2,3]}

In [561]:
df = pd.DataFrame(d)

In [562]:
df

A    B  C
0  1.0  5.0  1
1  2.0  NaN  2
2  NaN  NaN  3

- A lot of times you just wanna drop missing values
- To drop any row or column with missing values

In [564]:
df.dropna()

A    B  C
0  1.0  5.0  1

- Notice that above data is filtered from row '0'

- To drop missing values from the column, I will have to specify axis=1 with the bracket 

In [567]:
df.dropna(axis=1)

C
0  1
1  2
2  3

- Note that we can also specify a threshold
- Tips: Press 'Shift + tab' for suggestions

In [569]:
df.dropna(thresh=2)

A    B  C
0  1.0  5.0  1
1  2.0  NaN  2

- This drops the rows with at least 2 missing values 

## Alternative approach to fill the missing values

In [572]:
df.fillna(value='fillvalue')

A          B  C
0          1          5  1
1          2  fillvalue  2
2  fillvalue  fillvalue  3

- A lot of times we fill in the missing values with the mean and we can do that with the function 

In [574]:
df['A']

0    1.0
1    2.0
2    NaN
Name: A, dtype: float64

- Fill the Nan (missing values) with the mean value

In [576]:
df['A'].fillna(value=df['A'].mean())

0    1.0
1    2.0
2    1.5
Name: A, dtype: float64

In [577]:
df['B'].fillna(value=df['B'].mean())

0    5.0
1    5.0
2    5.0
Name: B, dtype: float64

In [578]:
df.fillna(value=5)

A    B  C
0  1.0  5.0  1
1  2.0  5.0  2
2  5.0  5.0  3

## Groupby or Aggregate Functions 

- A Groupby allows you to group a collection of data in a row or column and then aggregate them to get a single value based on mean or standard deviation.

In [581]:
import numpy as np
import pandas as pd

In [582]:
data = {'Company':['GOOG','GOOG','MSFT','MSFT','FB','FB'],
       'Person':['Sam','Charlie','Amy','Vanessa','Carl','Sarah'],
       'Sales':[200,120,340,124,234,350]}

In [583]:
df = pd.DataFrame(data)

In [584]:
df

Company   Person  Sales
0    GOOG      Sam    200
1    GOOG  Charlie    120
2    MSFT      Amy    340
3    MSFT  Vanessa    124
4      FB     Carl    234
5      FB    Sarah    350

In [585]:
df.groupby('Company')

In [586]:
byComp = df.groupby('Company')

In [587]:
byComp.mean()

Sales
Company       
FB         292
GOOG       160
MSFT       232

In [588]:
byComp.sum()

Sales
Company       
FB         584
GOOG       320
MSFT       464

In [589]:
byComp.std()

Sales
Company            
FB        82.024387
GOOG      56.568542
MSFT     152.735065

- **Call the sum of FB**

In [591]:
byComp.sum().loc['FB']

Sales    584
Name: FB, dtype: int64

- Presenting the data in one line 

In [593]:
df.groupby('Company').sum().loc['FB']

Sales    584
Name: FB, dtype: int64

In [594]:
df.groupby('Company').mean().loc['GOOG']

Sales    160
Name: GOOG, dtype: int64

- **Using other aggregate functuions such as count, max and min.**

In [596]:
df.groupby('Company').count()

Person  Sales
Company               
FB            2      2
GOOG          2      2
MSFT          2      2

In [597]:
df.groupby('Company').max()

Person  Sales
Company                
FB         Sarah    350
GOOG         Sam    200
MSFT     Vanessa    340

In [598]:
df.groupby('Company').min()

Person  Sales
Company                
FB          Carl    234
GOOG     Charlie    120
MSFT         Amy    124

## Describe method

In [600]:
df.groupby('Company').describe()

Sales                                                      
        count   mean         std    min    25%    50%    75%    max
Company                                                            
FB        2.0  292.0   82.024387  234.0  263.0  292.0  321.0  350.0
GOOG      2.0  160.0   56.568542  120.0  140.0  160.0  180.0  200.0
MSFT      2.0  232.0  152.735065  124.0  178.0  232.0  286.0  340.0

## Transpose

In [602]:
df.groupby('Company').describe().transpose()

Company              FB        GOOG        MSFT
Sales count    2.000000    2.000000    2.000000
      mean   292.000000  160.000000  232.000000
      std     82.024387   56.568542  152.735065
      min    234.000000  120.000000  124.000000
      25%    263.000000  140.000000  178.000000
      50%    292.000000  160.000000  232.000000
      75%    321.000000  180.000000  286.000000
      max    350.000000  200.000000  340.000000

- **Filter out the 'Facebook (FB)' company.**

In [604]:
df.groupby('Company').describe().loc['FB']

Sales  count      2.000000
       mean     292.000000
       std       82.024387
       min      234.000000
       25%      263.000000
       50%      292.000000
       75%      321.000000
       max      350.000000
Name: FB, dtype: float64

In [605]:
df.groupby('Company').describe().transpose()['GOOG']

Sales  count      2.000000
       mean     160.000000
       std       56.568542
       min      120.000000
       25%      140.000000
       50%      160.000000
       75%      180.000000
       max      200.000000
Name: GOOG, dtype: float64

## Task 1:

## Merging, Joining and Concatenating
- This functions link together the data in a chain or series in DataFrames

In [607]:
import pandas as pd

In [608]:
df1 = pd.DataFrame({'A':['A0','A1','A2','A3'],
                   'B':['B0','B1','B2','B3'],
                   'C':['C0','C1','C2','C3'],
                   'D':['D0','D1','D2','D3']},
                  index=[0,1,2,3])

In [609]:
df2 = pd.DataFrame({'A':['A4','A5','A6','A7'],
                   'B':['B4','B5','B6','B7'],
                   'C':['C4','C5','C6','C7'],
                   'D':['D4','D5','D6','D7']},
                   index=[4,5,6,7])

In [610]:
df3 = pd.DataFrame({'A':['A8','A9','A10','A11'],
                   'B':['B8','B9','B10','B11'],
                   'C':['C8','C9','C10','C11'],
                   'D':['D8','D9','D10','D11']},
                  index=[8,9,10,11])

In [611]:
df1

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2
3  A3  B3  C3  D3

In [612]:
df2

A   B   C   D
4  A4  B4  C4  D4
5  A5  B5  C5  D5
6  A6  B6  C6  D6
7  A7  B7  C7  D7

In [613]:
df3

A    B    C    D
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

- Now concatenate the df functions together
- Concatenation glues together DataFrames. Keep in mind that the dimensions should match along the axis you are concatenating on
- Use 'pd.concat' to pass in the list of DataFrames you want to concatenate together.

In [616]:
pd.concat([df1,df2,df3])

A    B    C    D
0    A0   B0   C0   D0
1    A1   B1   C1   D1
2    A2   B2   C2   D2
3    A3   B3   C3   D3
4    A4   B4   C4   D4
5    A5   B5   C5   D5
6    A6   B6   C6   D6
7    A7   B7   C7   D7
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

- Tips: Choosing your axis wisely, know your data before choosing your axis 

In [618]:
pd.concat([df1,df2,df3], axis=0)

A    B    C    D
0    A0   B0   C0   D0
1    A1   B1   C1   D1
2    A2   B2   C2   D2
3    A3   B3   C3   D3
4    A4   B4   C4   D4
5    A5   B5   C5   D5
6    A6   B6   C6   D6
7    A7   B7   C7   D7
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

In [619]:
pd.concat([df1,df2,df3],axis=1)

A    B    C    D    A    B    C    D    A    B    C    D
0    A0   B0   C0   D0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1    A1   B1   C1   D1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2    A2   B2   C2   D2  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
3    A3   B3   C3   D3  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
4   NaN  NaN  NaN  NaN   A4   B4   C4   D4  NaN  NaN  NaN  NaN
5   NaN  NaN  NaN  NaN   A5   B5   C5   D5  NaN  NaN  NaN  NaN
6   NaN  NaN  NaN  NaN   A6   B6   C6   D6  NaN  NaN  NaN  NaN
7   NaN  NaN  NaN  NaN   A7   B7   C7   D7  NaN  NaN  NaN  NaN
8   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   A8   B8   C8   D8
9   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   A9   B9   C9   D9
10  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  A10  B10  C10  D10
11  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  A11  B11  C11  D11

## Task 2: Concatenate DataFrame

In [621]:
left = pd.DataFrame({'Key':['K0','K1','K2','K3'],
                    'A':['A0','A1','A2','A3'],
                    'B':['B0','B1','B2','B3']})

right = pd.DataFrame({'Key':['K0','K1','K2','K3'],
                     'C':['C0','C1','C2','C3'],
                     'D':['D0','D1','D2','D3']})

In [622]:
left

Key   A   B
0  K0  A0  B0
1  K1  A1  B1
2  K2  A2  B2
3  K3  A3  B3

In [623]:
right

Key   C   D
0  K0  C0  D0
1  K1  C1  D1
2  K2  C2  D2
3  K3  C3  D3

- **Merge DataFrames together using similar logic as merging in SQL Tables.**

In [625]:
pd.merge(left,right,how='inner',on='Key')

Key   A   B   C   D
0  K0  A0  B0  C0  D0
1  K1  A1  B1  C1  D1
2  K2  A2  B2  C2  D2
3  K3  A3  B3  C3  D3

# Alternative Method

In [627]:
left = pd.DataFrame({'Key1':['K0','K0','K1','K2'],
                    'Key2':['K0','K1','K0','K1'],
                    'A':['A0','A1','A2','A3'],
                    'B':['B0','B1','B2','B3']})

right = pd.DataFrame({'Key1':['K0','K1','K1','K2'],
                     'Key2':['K0','K0','K0','K0'],
                     'C':['C0','C1','C2','C3'],
                     'D':['D0','D1','D2','D3']})

In [628]:
pd.merge(left,right,on=['Key1','Key2'])

Key1 Key2   A   B   C   D
0   K0   K0  A0  B0  C0  D0
1   K1   K0  A2  B2  C1  D1
2   K1   K0  A2  B2  C2  D2

- Uisng outer merging approach

In [630]:
pd.merge(left,right, how='outer', on=['Key1','Key2'])

Key1 Key2    A    B    C    D
0   K0   K0   A0   B0   C0   D0
1   K0   K1   A1   B1  NaN  NaN
2   K1   K0   A2   B2   C1   D1
3   K1   K0   A2   B2   C2   D2
4   K2   K1   A3   B3  NaN  NaN
5   K2   K0  NaN  NaN   C3   D3

In [631]:
pd.merge(left,right, how='right',on=['Key1','Key2'])

Key1 Key2    A    B   C   D
0   K0   K0   A0   B0  C0  D0
1   K1   K0   A2   B2  C1  D1
2   K1   K0   A2   B2  C2  D2
3   K2   K0  NaN  NaN  C3  D3

In [632]:
pd.merge(left,right,how='left',on=['Key1','Key2'])

Key1 Key2   A   B    C    D
0   K0   K0  A0  B0   C0   D0
1   K0   K1  A1  B1  NaN  NaN
2   K1   K0  A2  B2   C1   D1
3   K1   K0  A2  B2   C2   D2
4   K2   K1  A3  B3  NaN  NaN

## Joining DataFrames

- This is a convenient method for combining the columns of two potentially differntly-indexed DataFrames into a single DataFrame

In [634]:
left = pd.DataFrame({'A':['A0','A1','A2'],
                    'B':['B0','B1','B2']},
                   index=['K0','K1','K2'])

right = pd.DataFrame({'C':['C0','C2','C3'],
                     'D':['D0','D2','D3']},
                    index=['K0','K2','K3'])

- Join the DataFrames

In [636]:
left.join(right)

A   B    C    D
K0  A0  B0   C0   D0
K1  A1  B1  NaN  NaN
K2  A2  B2   C2   D2

In [637]:
left.join(right, how='outer')

A    B    C    D
K0   A0   B0   C0   D0
K1   A1   B1  NaN  NaN
K2   A2   B2   C2   D2
K3  NaN  NaN   C3   D3

- **Some useful operations.**

In [639]:
import numpy as np 
import pandas as pd

In [640]:
df = pd.DataFrame({'Col1':[1,2,3,4],
                  'Col2':[444,555,666,444],
                  'Col3':['abc','def','ghi','xyz']})

df.head()

Col1  Col2 Col3
0     1   444  abc
1     2   555  def
2     3   666  ghi
3     4   444  xyz

### Finding a unique value in a DataFrame

- There are 3 approached for finding the unique values in DataFrames
- For example find the unique value in 'Col2'

In [643]:
df['Col2'].unique()

array([444, 555, 666])

- **Find the length of Values in Col2.**

In [645]:
len(df['Col2'].unique())

3

- **Using Pandas function.**

In [647]:
df['Col2'].nunique()

3

## Value Count

- Checking the number of times a number appears in a sample of data 
 

In [649]:
df['Col2'].value_counts()

444    2
555    1
666    1
Name: Col2, dtype: int64

- We can see that the number '444' appeared twice the sample data.

- Selecting Data

In [652]:
df[df['Col1']>2]

Col1  Col2 Col3
2     3   666  ghi
3     4   444  xyz

In [653]:
df['Col1']>2

0    False
1    False
2     True
3     True
Name: Col1, dtype: bool

### Applying Conditional Selection 

In [655]:
df[(df['Col1']>2) & (df['Col2']==444)]

Col1  Col2 Col3
3     4   444  xyz

- **To demonstrate the above method:**
- Imagine we have a function 

In [657]:
def times2(x):
    return x*2

In [658]:
df['Col1'].sum()

10

In [659]:
df['Col1'].apply(times2)

0    2
1    4
2    6
3    8
Name: Col1, dtype: int64

In [660]:
df['Col3']

0    abc
1    def
2    ghi
3    xyz
Name: Col3, dtype: object

- **Find the lenght of the strings.**

In [662]:
df['Col3'].apply(len)

0    3
1    3
2    3
3    3
Name: Col3, dtype: int64

- **Tips: apply is a powerful tool when combined with lambda expression.**
- Using 'lambda expression'

In [665]:
df['Col2'].apply(lambda x: x*2)

0     888
1    1110
2    1332
3     888
Name: Col2, dtype: int64

- **it is a very powerful, when you have the ability to apply your own custom lambda expressions.**

In [667]:
df

Col1  Col2 Col3
0     1   444  abc
1     2   555  def
2     3   666  ghi
3     4   444  xyz

- Romoving column 'Col3'

In [669]:
df.drop('Col3', axis=1)

Col1  Col2
0     1   444
1     2   555
2     3   666
3     4   444

- **Get the column names in the DataFrame.**

In [671]:
df.columns

Index(['Col1', 'Col2', 'Col3'], dtype='object')

In [672]:
df.index

RangeIndex(start=0, stop=4, step=1)

## Sorting and Ordering a DataFrame.
- Sort using 'Col2'

In [674]:
df.sort_values('Col2')

Col1  Col2 Col3
0     1   444  abc
3     4   444  xyz
1     2   555  def
2     3   666  ghi

- Notice how the index stays attached to the data even after they have been sorted so that you don't loose the data information 

- **Another very useful operation to find nul values in DataFrames in the .isnull()**

In [677]:
df.isnull()

Col1   Col2   Col3
0  False  False  False
1  False  False  False
2  False  False  False
3  False  False  False

- Notice all the values in the column is false becos we have no null value

## The Pivot Table in Python
- This is similar to pivot table advanced Excel 

In [680]:
data = {'A':['foo','foo','foo','bar','bar','bar'],
       'B':['one','one','two','two','one','one'],
       'C':['X','Y','X','Y','X','Y'],
       'D':[1,3,2,5,4,1]}

df = pd.DataFrame(data)

In [681]:
df

A    B  C  D
0  foo  one  X  1
1  foo  one  Y  3
2  foo  two  X  2
3  bar  two  Y  5
4  bar  one  X  4
5  bar  one  Y  1

- **Create a Multi-Index pivot table**

In [683]:
df.pivot_table(values='D', index=['A','B'],columns=['C'])

C          X    Y
A   B            
bar one  4.0  1.0
    two  NaN  5.0
foo one  1.0  3.0
    two  2.0  NaN

## Data Input and Output

- The four main data input file format considered: 'CSV'; 'EXCEL'; 'HTML'; 'SQL'

- **In order to be able to work with SQL or HTML file**, I will install the following libraries:
- `conda install sqlalchemy` or `pip install sqlalchemy`
- `conda install lxml` or `pip install lxml`
- `conda install html5lib` or `pip install html5lib`
- `conda install BeautifulSoup4` or `pip install BeautifulSoup4`

In [688]:
import pandas as pd

## Open and Read CSV file

- First I will check for the path of the file using 'pwd`

In [2]:
#pwd

In [745]:
pd.read_csv('example.csv')

B;C;D;E
0      a;b;c;d
1      0;1;2;3
2      4;5;6;7
3    8;9;10;11
4  12;13;14;15

In [746]:
pd.read_excel('Excel_Sample.xlsx')

Unnamed: 0   a   b   c   d
0           0   0   1   2   3
1           1   4   5   6   7
2           2   8   9  10  11
3           3  12  13  14  15

In [747]:
df = pd.read_csv('example.csv')

In [695]:
df

A    B  C  D
0  foo  one  X  1
1  foo  one  Y  3
2  foo  two  X  2
3  bar  two  Y  5
4  bar  one  X  4
5  bar  one  Y  1

In [748]:
df = pd.read_excel('Excel_Sample.xlsx')

In [697]:
df

A    B  C  D
0  foo  one  X  1
1  foo  one  Y  3
2  foo  two  X  2
3  bar  two  Y  5
4  bar  one  X  4
5  bar  one  Y  1

## Writing to the CSV file

- Use df.to_(plus tab to get more suggestions)

In [699]:
df.to_csv('My_output')

In [700]:
pd.read_csv('My_output')

Unnamed: 0    A    B  C  D
0           0  foo  one  X  1
1           1  foo  one  Y  3
2           2  foo  two  X  2
3           3  bar  two  Y  5
4           4  bar  one  X  4
5           5  bar  one  Y  1

- Correct the double indexing by setting index=False

In [702]:
df.to_csv('My_output', index=False)

In [703]:
pd.read_csv('My_output')

A    B  C  D
0  foo  one  X  1
1  foo  one  Y  3
2  foo  two  X  2
3  bar  two  Y  5
4  bar  one  X  4
5  bar  one  Y  1

## Reading and Writing to Excel file

- install xlrd using `conda install xlrd`

In [749]:
pd.read_excel('Excel_Sample.xlsx',Sheetname='Sheet1')

Unnamed: 0   a   b   c   d
0           0   0   1   2   3
1           1   4   5   6   7
2           2   8   9  10  11
3           3  12  13  14  15

- Notice tha the Sheetname and Sheet1 both begin with capital 'S'

- But to output or save the data to excel with a new name, the sheet_name begins with small letter 's'

In [711]:
df.to_excel('Excel_Sample2.xlsx', sheet_name='NewSheet')

- A new Excel sheet called Excel_Sample2 have been created in the folder

## Reading Data file From HTML

- To read data from an HTML file/website

In [714]:
data = pd.read_html('https://www.fdic.gov/bank/individual/failed/banklist.html')

In [715]:
type(data)

list

- Using pandas make a list of each iterm and convert them into a DataFrame
- Next, you would want to cycle through the list until you can locate exactly what you are looking for

In [718]:
data[0]

Bank Name           City  ST   CERT  \
0                 The First State Bank  Barboursville  WV  14361   
1                   Ericson State Bank        Ericson  NE  18265   
2     City National Bank of New Jersey         Newark  NJ  21111   
3                        Resolute Bank         Maumee  OH  58317   
4                Louisa Community Bank         Louisa  KY  58112   
..                                 ...            ...  ..    ...   
556                 Superior Bank, FSB       Hinsdale  IL  32646   
557                Malta National Bank          Malta  OH   6629   
558    First Alliance Bank & Trust Co.     Manchester  NH  34264   
559  National State Bank of Metropolis     Metropolis  IL   3815   
560                   Bank of Honolulu       Honolulu  HI  21029   

                   Acquiring Institution       Closing Date  
0                         MVB Bank, Inc.      April 3, 2020  
1             Farmers and Merchants Bank  February 14, 2020  
2                        Industrial Bank   November 1, 2019  
3                     Buckeye State Bank   October 25, 2019  
4      Kentucky Farmers Bank Corporation   October 25, 2019  
..                                   ...                ...  
556                Superior Federal, FSB      July 27, 2001  
557                    North Valley Bank        May 3, 2001  
558  Southern New Hampshire Bank & Trust   February 2, 2001  
559              Banterra Bank of Marion  December 14, 2000  
560                   Bank of the Orient   October 13, 2000  

[561 rows x 6 columns]

- Notice that the information at zero is the Bank name, city etc: As can be seen at the headings

In [720]:
data[0].head()

Bank Name           City  ST   CERT  \
0              The First State Bank  Barboursville  WV  14361   
1                Ericson State Bank        Ericson  NE  18265   
2  City National Bank of New Jersey         Newark  NJ  21111   
3                     Resolute Bank         Maumee  OH  58317   
4             Louisa Community Bank         Louisa  KY  58112   

               Acquiring Institution       Closing Date  
0                     MVB Bank, Inc.      April 3, 2020  
1         Farmers and Merchants Bank  February 14, 2020  
2                    Industrial Bank   November 1, 2019  
3                 Buckeye State Bank   October 25, 2019  
4  Kentucky Farmers Bank Corporation   October 25, 2019

## Reading Data From SQL
- Pandas isn't the best engine for reading SQL file because there are many flavours of SQL engines such as MySQL, SQL and PostgreSQL

- **Building a simple SQL Engine.**
- **A step by step approach to build a simple SQL engine memory.**

In [725]:
from sqlalchemy import create_engine

In [726]:
engine = create_engine('sqlite:///:memory:')

In [727]:
df.to_sql('my_table',engine)

In [728]:
sqldf = pd.read_sql('my_table',con=engine)

In [729]:
sqldf

index    A    B  C  D
0      0  foo  one  X  1
1      1  foo  one  Y  3
2      2  foo  two  X  2
3      3  bar  two  Y  5
4      4  bar  one  X  4
5      5  bar  one  Y  1

- `con=engine` **is a connection allowing you to pass the table to the sql engine 'sqlite:///:memory'**

## Data Analysis Project 1

In [741]:
import pandas as pd

- Read Salaries.csv as a dataframe called sal

In [750]:
sal = pd.read_csv('Salaries.csv')

- Check the head of the DataFrame 

In [751]:
sal.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.00  400184.25       NaN  567595.43         567595.43   
1  155966.02    245131.88  137811.38       NaN  538909.28         538909.28   
2  212739.13    106088.18   16452.60       NaN  335279.91         335279.91   
3   77916.00     56120.71  198306.90       NaN  332343.61         332343.61   
4  134401.60      9737.00  182234.59       NaN  326373.19         326373.19   

   Year  Notes         Agency  Status  
0  2011    NaN  San Francisco     NaN  
1  2011    NaN  San Francisco     NaN  
2  2011    NaN  San Francisco     NaN  
3  2011    NaN  San Francisco     NaN  
4  2011    NaN  San Francisco     NaN

- Make the DataFrame 2 rows 

In [ ]:
sal.head(2)

- Use the .info() method to find out how many entries in the DataFrame 

In [12]:
sal.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
Id                  148654 non-null int64
EmployeeName        148654 non-null object
JobTitle            148654 non-null object
BasePay             148045 non-null float64
OvertimePay         148650 non-null float64
OtherPay            148650 non-null float64
Benefits            112491 non-null float64
TotalPay            148654 non-null float64
TotalPayBenefits    148654 non-null float64
Year                148654 non-null int64
Notes               0 non-null float64
Agency              148654 non-null object
Status              0 non-null float64
dtypes: float64(8), int64(2), object(3)
memory usage: 14.7+ MB


- **Finding the average BasePay**

In [14]:
sal['BasePay'].mean()

66325.44884050643

- What is the highest amount of OvertimePay in the database?

In [17]:
sal['OvertimePay'].max()

245131.88

- What is the job title of JOSEPH DRISCOLL? 
- Note: Use all caps, otherwise you may get an answer that doesn't match up (there is also a lowercase Joseph Driscoll)

In [20]:
sal[sal['EmployeeName']=='JOSEPH DRISCOLL']['JobTitle']

24    CAPTAIN, FIRE SUPPRESSION
Name: JobTitle, dtype: object

- How much does JOSEPH DRISCOLL make (including benefits)?

In [23]:
sal[sal['EmployeeName']=='JOSEPH DRISCOLL']['TotalPayBenefits']

24    270324.91
Name: TotalPayBenefits, dtype: float64

- What is the name of highest paid person (including benefits)?

In [29]:
sal['TotalPayBenefits'].max()

567595.43

- But I want to know the name of the person receiving this amount 

In [31]:
sal['TotalPayBenefits']==sal['TotalPayBenefits'].max()

0          True
1         False
2         False
3         False
4         False
          ...  
148649    False
148650    False
148651    False
148652    False
148653    False
Name: TotalPayBenefits, Length: 148654, dtype: bool

- **Notice that the first person is true** 

In [32]:
sal[sal['TotalPayBenefits']==sal['TotalPayBenefits'].max()]

Id    EmployeeName                                        JobTitle  \
0   1  NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18          0.0  400184.25       NaN  567595.43         567595.43   

   Year  Notes         Agency  Status  
0  2011    NaN  San Francisco     NaN

- Now to grab his name

In [27]:
sal[sal['TotalPayBenefits']==sal['TotalPayBenefits'].max()]['EmployeeName']

0    NATHANIEL FORD
Name: EmployeeName, dtype: object

## Alternative method 

In [35]:
sal.loc[sal['TotalPayBenefits'].idxmax()]

Id                                                               1
EmployeeName                                        NATHANIEL FORD
JobTitle            GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY
BasePay                                                     167411
OvertimePay                                                      0
OtherPay                                                    400184
Benefits                                                       NaN
TotalPay                                                    567595
TotalPayBenefits                                            567595
Year                                                          2011
Notes                                                          NaN
Agency                                               San Francisco
Status                                                         NaN
Name: 0, dtype: object

- Using `sal.loc[]`

In [38]:
#sal.loc[]
sal['TotalPayBenefits'].idxmax()

0

In [39]:
sal.iloc[sal['TotalPayBenefits'].idxmax()]

Id                                                               1
EmployeeName                                        NATHANIEL FORD
JobTitle            GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY
BasePay                                                     167411
OvertimePay                                                      0
OtherPay                                                    400184
Benefits                                                       NaN
TotalPay                                                    567595
TotalPayBenefits                                            567595
Year                                                          2011
Notes                                                          NaN
Agency                                               San Francisco
Status                                                         NaN
Name: 0, dtype: object

- What is the name of the lowest paid person (including benefits)? Do you notice something strange about how much he or she is paid?

In [34]:
sal.loc[sal['TotalPayBenefits'].idxmin()]

Id                                      148654
EmployeeName                         Joe Lopez
JobTitle            Counselor, Log Cabin Ranch
BasePay                                      0
OvertimePay                                  0
OtherPay                               -618.13
Benefits                                     0
TotalPay                               -618.13
TotalPayBenefits                       -618.13
Year                                      2014
Notes                                      NaN
Agency                           San Francisco
Status                                     NaN
Name: 148653, dtype: object

In [41]:
sal[sal['TotalPayBenefits']==sal['TotalPayBenefits'].min()]['EmployeeName']

148653    Joe Lopez
Name: EmployeeName, dtype: object

- What was the average (mean) BasePay of all employees per year? (2011-2014)?

In [45]:
sal.groupby('Year').mean()

Id       BasePay  OvertimePay     OtherPay      Benefits  \
Year                                                                   
2011   18080.0  63595.956517  4531.065429  3617.081926           NaN   
2012   54542.5  65436.406857  5023.417824  3653.437583  26439.966967   
2013   91728.5  69630.030216  5281.641980  3819.969007  23829.076572   
2014  129593.0  66564.421924  5401.993737  3505.421251  24789.601756   

          TotalPay  TotalPayBenefits  Notes  Status  
Year                                                 
2011  71744.103871      71744.103871    NaN     NaN  
2012  74113.262265     100553.229232    NaN     NaN  
2013  77611.443142     101440.519714    NaN     NaN  
2014  75463.918140     100250.918884    NaN     NaN

In [44]:
sal.groupby('Year').mean()['BasePay']

Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64

- How many unique job titles are there?

In [47]:
sal['JobTitle'].nunique()

2159

## Alternative method
- using `unique` instead of `nunique`

In [50]:
len(sal['JobTitle'].unique())

2159

- What are the top 5 most common jobs?

In [52]:
sal['JobTitle'].value_counts().head(5)

Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: JobTitle, dtype: int64

- How many Job Titles were represented by only one person in 2013? (e.g. Job Titles with only one occurence in 2013?) 

In [54]:
sal[sal['Year']==2013]['JobTitle'].value_counts()

Transit Operator                2295
Special Nurse                   1445
Registered Nurse                1231
Public Svc Aide-Public Works     859
Custodian                        789
                                ... 
Employment & Training Spec 6       1
Shade and Drapery Worker           1
City Attorney                      1
Mayoral Staff X                    1
Sr Employee Asst Counselor         1
Name: JobTitle, Length: 1051, dtype: int64

In [55]:
sal[sal['Year']==2013]['JobTitle'].value_counts()==1

Transit Operator                False
Special Nurse                   False
Registered Nurse                False
Public Svc Aide-Public Works    False
Custodian                       False
                                ...  
Employment & Training Spec 6     True
Shade and Drapery Worker         True
City Attorney                    True
Mayoral Staff X                  True
Sr Employee Asst Counselor       True
Name: JobTitle, Length: 1051, dtype: bool

In [57]:
sum(sal[sal['Year']==2013]['JobTitle'].value_counts()==1)

202

- How many people have the word Chief in their job title? (This is pretty tricky)
- In order to solve this task, I will have to create a custome function

In [60]:
sal['JobTitle']

0         GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY
1                        CAPTAIN III (POLICE DEPARTMENT)
2                        CAPTAIN III (POLICE DEPARTMENT)
3                   WIRE ROPE CABLE MAINTENANCE MECHANIC
4           DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)
                               ...                      
148649                                         Custodian
148650                                      Not provided
148651                                      Not provided
148652                                      Not provided
148653                        Counselor, Log Cabin Ranch
Name: JobTitle, Length: 148654, dtype: object

In [63]:
sal['JobTitle'].iloc[0]

'GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY'

- Now lets create a function 

In [62]:
def chief_string(title):
    if 'chief' in title.lower().split():
        return True
    else:
        return False

In [64]:
chief_string('GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY')

False

- This 'False' makes sense because there isnt any Chief in the job title. If we change the General to Chief

In [66]:
chief_string('CHIEF MANAGER-METROPOLITAN TRANSIT AUTHORITY')

True

In [67]:
sal['JobTitle'].apply(lambda x: chief_string(x))

0         False
1         False
2         False
3         False
4          True
          ...  
148649    False
148650    False
148651    False
148652    False
148653    False
Name: JobTitle, Length: 148654, dtype: bool

In [68]:
sum(sal['JobTitle'].apply(lambda x: chief_string(x)))

477

- 477 people have the word chief in their title 

- Find if there is a correlation between length of the Job Title string and Salary?

In [71]:
sal['title_len']= sal['JobTitle'].apply(len)

In [72]:
sal[['JobTitle', 'title_len']]

JobTitle  title_len
0       GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY         46
1                      CAPTAIN III (POLICE DEPARTMENT)         31
2                      CAPTAIN III (POLICE DEPARTMENT)         31
3                 WIRE ROPE CABLE MAINTENANCE MECHANIC         36
4         DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)         44
...                                                ...        ...
148649                                       Custodian          9
148650                                    Not provided         12
148651                                    Not provided         12
148652                                    Not provided         12
148653                      Counselor, Log Cabin Ranch         26

[148654 rows x 2 columns]

In [73]:
sal[['JobTitle','title_len']].corr()

title_len
title_len        1.0

- But I want correlation between the JobTitle and Salary 

In [75]:
sal[['TotalPayBenefits','title_len']].corr()

TotalPayBenefits  title_len
TotalPayBenefits          1.000000  -0.036878
title_len                -0.036878   1.000000

- Here we see that there is no correlation on how long JobTitle and TotalPayBenefits

## Data Analysis Project 2

In [79]:
import pandas as pd

In [80]:
ecom = pd.read_csv('Ecommerce Purchases')

In [81]:
ecom.head()

Address    Lot AM or PM  \
0  16629 Pace Camp Apt. 448\nAlexisborough, NE 77...  46 in       PM   
1  9374 Jasmine Spurs Suite 508\nSouth John, TN 8...  28 rn       PM   
2                   Unit 0065 Box 5052\nDPO AP 27450  94 vE       PM   
3              7780 Julia Fords\nNew Stacy, WA 45798  36 vm       PM   
4  23012 Munoz Drive Suite 337\nNew Cynthia, TX 5...  20 IE       AM   

                                        Browser Info  \
0  Opera/9.56.(X11; Linux x86_64; sl-SI) Presto/2...   
1  Opera/8.93.(Windows 98; Win 9x 4.90; en-US) Pr...   
2  Mozilla/5.0 (compatible; MSIE 9.0; Windows NT ...   
3  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_0 ...   
4  Opera/9.58.(X11; Linux x86_64; it-IT) Presto/2...   

                           Company       Credit Card CC Exp Date  \
0                  Martinez-Herman  6011929061123406       02/20   
1  Fletcher, Richards and Whitaker  3337758169645356       11/18   
2       Simpson, Williams and Pham      675957666125       08/19   
3  Williams, Marshall and Buchanan  6011578504430710       02/24   
4        Brown, Watson and Andrews  6011456623207998       10/25   

   CC Security Code                  CC Provider  \
0               900                 JCB 16 digit   
1               561                   Mastercard   
2               699                 JCB 16 digit   
3               384                     Discover   
4               678  Diners Club / Carte Blanche   

                            Email                                     Job  \
0               pdunlap@yahoo.com  Scientist, product/process development   
1              anthony41@reed.com                       Drilling engineer   
2  amymiller@morales-harrison.com                Customer service manager   
3     brent16@olson-robinson.info                       Drilling engineer   
4     christopherwright@gmail.com                             Fine artist   

        IP Address Language  Purchase Price  
0  149.146.147.205       el           98.14  
1     15.160.41.51       fr           70.73  
2   132.207.160.22       de            0.95  
3     30.250.74.19       es           78.04  
4     24.140.33.94       es           77.82

- How many rows and columns are there?

In [84]:
ecom.columns

Index(['Address', 'Lot', 'AM or PM', 'Browser Info', 'Company', 'Credit Card',
       'CC Exp Date', 'CC Security Code', 'CC Provider', 'Email', 'Job',
       'IP Address', 'Language', 'Purchase Price'],
      dtype='object')

- Show the lenght

In [85]:
len(ecom.columns)

14

- To show how many rows in the sample data

In [87]:
len(ecom.index)

10000

In [88]:
ecom.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
Address             10000 non-null object
Lot                 10000 non-null object
AM or PM            10000 non-null object
Browser Info        10000 non-null object
Company             10000 non-null object
Credit Card         10000 non-null int64
CC Exp Date         10000 non-null object
CC Security Code    10000 non-null int64
CC Provider         10000 non-null object
Email               10000 non-null object
Job                 10000 non-null object
IP Address          10000 non-null object
Language            10000 non-null object
Purchase Price      10000 non-null float64
dtypes: float64(1), int64(2), object(11)
memory usage: 1.1+ MB


- What is the average Purchase Price? Notice from the column we have 'Purchase Price'

In [90]:
ecom['Purchase Price'].mean()

50.34730200000025

- What were the highest and lowest purchase prices? 

In [92]:
ecom['Purchase Price'].max()

99.99

In [93]:
ecom['Purchase Price'].min()

0.0

- How many people have English 'en' as their Language of choice on the website?

In [95]:
ecom.head(3)

Address    Lot AM or PM  \
0  16629 Pace Camp Apt. 448\nAlexisborough, NE 77...  46 in       PM   
1  9374 Jasmine Spurs Suite 508\nSouth John, TN 8...  28 rn       PM   
2                   Unit 0065 Box 5052\nDPO AP 27450  94 vE       PM   

                                        Browser Info  \
0  Opera/9.56.(X11; Linux x86_64; sl-SI) Presto/2...   
1  Opera/8.93.(Windows 98; Win 9x 4.90; en-US) Pr...   
2  Mozilla/5.0 (compatible; MSIE 9.0; Windows NT ...   

                           Company       Credit Card CC Exp Date  \
0                  Martinez-Herman  6011929061123406       02/20   
1  Fletcher, Richards and Whitaker  3337758169645356       11/18   
2       Simpson, Williams and Pham      675957666125       08/19   

   CC Security Code   CC Provider                           Email  \
0               900  JCB 16 digit               pdunlap@yahoo.com   
1               561    Mastercard              anthony41@reed.com   
2               699  JCB 16 digit  amymiller@morales-harrison.com   

                                      Job       IP Address Language  \
0  Scientist, product/process development  149.146.147.205       el   
1                       Drilling engineer     15.160.41.51       fr   
2                Customer service manager   132.207.160.22       de   

   Purchase Price  
0           98.14  
1           70.73  
2            0.95

- This shows how many strings in the language column 

In [97]:
ecom['Language']=='en'

0       False
1       False
2       False
3       False
4       False
        ...  
9995    False
9996    False
9997    False
9998    False
9999    False
Name: Language, Length: 10000, dtype: bool

In [98]:
ecom[ecom['Language']=='en']

Address    Lot AM or PM  \
11               349 Laurie Parks\nThomasview, ID 08970  30 kK       PM   
12      733 Heather Rest Apt. 670\nBoltonport, UT 78662  69 DO       AM   
22     625 Laura Summit Suite 112\nShortmouth, AK 97156  45 oa       PM   
26          5334 Sheppard Fort\nNorth Kirsten, ND 71170  40 ND       AM   
31                     USNS Alvarado\nFPO AA 27052-1231  26 Lh       PM   
...                                                 ...    ...      ...   
9961        714 Joshua Ridges\nCarpenterville, NJ 37418  84 rK       PM   
9963  3865 Davis Meadow Suite 915\nPort Stacieview, ...  11 fo       AM   
9970              0060 Keith Stream\nWestport, CO 47097  11 nt       PM   
9973     92126 Daniel Streets\nNorth Patricia, IA 10477  04 QN       PM   
9979  00490 Joseph Haven Apt. 149\nWest Keithfurt, S...  69 XK       PM   

                                           Browser Info  \
11    Mozilla/5.0 (X11; Linux i686; rv:1.9.6.20) Gec...   
12    Mozilla/5.0 (Macintosh; Intel Mac OS X 10_5_3 ...   
22    Mozilla/5.0 (Windows 95) AppleWebKit/5342 (KHT...   
26    Mozilla/5.0 (Macintosh; Intel Mac OS X 10_6_3)...   
31    Opera/8.84.(X11; Linux i686; sl-SI) Presto/2.9...   
...                                                 ...   
9961  Opera/9.49.(X11; Linux i686; it-IT) Presto/2.9...   
9963  Mozilla/5.0 (iPod; U; CPU iPhone OS 3_3 like M...   
9970  Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10_8...   
9973  Mozilla/5.0 (Windows NT 5.2) AppleWebKit/5362 ...   
9979  Mozilla/5.0 (Windows NT 6.1) AppleWebKit/5311 ...   

                          Company       Credit Card CC Exp Date  \
11                     Kim-Oliver   869975209012056       06/26   
12                   Moore-Martin  5115990487067905       05/26   
22                    Burnett Ltd  3158663456056588       07/22   
26    Stephens, Johnson and Combs       60488441233       03/19   
31                Nicholson Group     4614997834548       03/22   
...                           ...               ...         ...   
9961                 White-Wilson  3158062139581601       08/16   
9963      Brown, Gentry and White  4295151255424068       07/25   
9970     Carpenter, Good and Hart  6011485664704662       07/19   
9973                   Morrow Inc      501850922854       02/19   
9979              Randall-Ramirez   346734838013255       10/23   

      CC Security Code       CC Provider                       Email  \
11                9717      JCB 15 digit    johnnymiller@coleman.com   
12                 119     VISA 16 digit           tholt@hotmail.com   
22                 508          Discover   chambersmichael@gmail.com   
26                  70     VISA 13 digit            mlogan@gmail.com   
31                 909        Mastercard        ashley12@hotmail.com   
...                ...               ...                         ...   
9961               723     VISA 13 digit  justin11@porter-smith.info   
9963               764      JCB 15 digit            zjames@gmail.com   
9970               543          Discover     rangelbrian@hotmail.com   
9973              2708  American Express            rlewis@yahoo.com   
9979               224      JCB 16 digit          qjohnson@yahoo.com   

                            Job      IP Address Language  Purchase Price  
11      Diagnostic radiographer  128.222.40.234       en           19.26  
12           Surveyor, quantity  236.71.234.240       en           39.65  
22               Retail manager   65.106.61.141       en           33.62  
26            Barrister's clerk  61.197.134.185       en           17.75  
31              Sales executive  94.176.142.201       en           94.14  
...                         ...             ...      ...             ...  
9961       Nurse, mental health   40.218.118.68       en           23.96  
9963                    Surgeon  171.24.198.172       en           58.17  
9970        Electrical engineer    242.8.85.205       en           17.76  
9973  Editor, maga

- Now to check for the total number 

In [100]:
ecom[ecom['Language']=='en'].count()

Address             1098
Lot                 1098
AM or PM            1098
Browser Info        1098
Company             1098
Credit Card         1098
CC Exp Date         1098
CC Security Code    1098
CC Provider         1098
Email               1098
Job                 1098
IP Address          1098
Language            1098
Purchase Price      1098
dtype: int64

- To be specific 

In [102]:
ecom[ecom['Language']=='en']['Language'].count()

1098

- How many people have the job title of "Lawyer" ? 

In [104]:
ecom['Job']=='Lawyer'

0       False
1       False
2       False
3       False
4       False
        ...  
9995    False
9996    False
9997    False
9998    False
9999    False
Name: Job, Length: 10000, dtype: bool

In [105]:
ecom[ecom['Job']=='Lawyer'].count()

Address             30
Lot                 30
AM or PM            30
Browser Info        30
Company             30
Credit Card         30
CC Exp Date         30
CC Security Code    30
CC Provider         30
Email               30
Job                 30
IP Address          30
Language            30
Purchase Price      30
dtype: int64

In [108]:
ecom[ecom['Job']=='Lawyer'].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30 entries, 470 to 9979
Data columns (total 14 columns):
Address             30 non-null object
Lot                 30 non-null object
AM or PM            30 non-null object
Browser Info        30 non-null object
Company             30 non-null object
Credit Card         30 non-null int64
CC Exp Date         30 non-null object
CC Security Code    30 non-null int64
CC Provider         30 non-null object
Email               30 non-null object
Job                 30 non-null object
IP Address          30 non-null object
Language            30 non-null object
Purchase Price      30 non-null float64
dtypes: float64(1), int64(2), object(11)
memory usage: 3.5+ KB


In [109]:
len(ecom[ecom['Job']=='Lawyer'].index)

30

- How many people made the purchase during the AM and how many people made the purchase during PM ?
`(Hint: Check out value_counts() ` 

In [111]:
ecom['AM or PM'].unique

<bound method Series.unique of 0       PM
1       PM
2       PM
3       PM
4       AM
        ..
9995    PM
9996    AM
9997    AM
9998    PM
9999    AM
Name: AM or PM, Length: 10000, dtype: object>

In [112]:
ecom['AM or PM'].value_counts()

PM    5068
AM    4932
Name: AM or PM, dtype: int64

- What are the 5 most common Job Titles? 

In [114]:
ecom['Job']

0       Scientist, product/process development
1                            Drilling engineer
2                     Customer service manager
3                            Drilling engineer
4                                  Fine artist
                         ...                  
9995                                Printmaker
9996                           Energy engineer
9997                        Veterinary surgeon
9998                  Local government officer
9999                    Embryologist, clinical
Name: Job, Length: 10000, dtype: object

In [115]:
ecom['Job'].value_counts()

Interior and spatial designer        31
Lawyer                               30
Social researcher                    28
Purchasing manager                   27
Research officer, political party    27
                                     ..
Stage manager                         7
Actuary                               6
Editor, film/video                    6
Investment analyst                    5
Information systems manager           5
Name: Job, Length: 623, dtype: int64

- Now top 5 most common Job Titles

In [117]:
ecom['Job'].value_counts().head(5)

Interior and spatial designer        31
Lawyer                               30
Social researcher                    28
Purchasing manager                   27
Research officer, political party    27
Name: Job, dtype: int64

- Someone made a purchase that came from Lot: "90 WT" , what was the Purchase Price for this transaction? 

In [126]:
ecom['Lot']=='90 WT'

0       False
1       False
2       False
3       False
4       False
        ...  
9995    False
9996    False
9997    False
9998    False
9999    False
Name: Lot, Length: 10000, dtype: bool

In [127]:
ecom[ecom['Lot']=='90 WT']

Address    Lot AM or PM  \
513  50398 Mccoy Rest Suite 597\nSouth Garyborough,...  90 WT       AM   

                                          Browser Info     Company  \
513  Mozilla/5.0 (iPod; U; CPU iPhone OS 3_2 like M...  Bright PLC   

      Credit Card CC Exp Date  CC Security Code       CC Provider  \
513  630438419693       11/19               173  American Express   

                Email              Job     IP Address Language  Purchase Price  
513  jesse00@page.net  Energy engineer  156.70.208.94       ru            75.1

In [128]:
ecom[ecom['Lot']=='90 WT']['Purchase Price']

513    75.1
Name: Purchase Price, dtype: float64

- What is the email of the person with the following Credit Card Number: 4926535242672853 (please note that the information are made up and not real)

In [132]:
ecom[ecom['Credit Card']==4926535242672853]['Email']

1234    bondellen@williams-garza.com
Name: Email, dtype: object

- How many people have American Express as their Credit Card Provider and made a purchase above $95?

In [138]:
ecom[(ecom['CC Provider']=='American Express') & (ecom['Purchase Price']>95)]

Address    Lot AM or PM  \
9       3795 Dawson Extensions\nLake Tinafort, ID 88739  15 Ug       AM   
280   81060 Dustin Causeway Apt. 503\nPort Danielche...  80 zh       PM   
372              359 Stanley Coves\nSalasfort, SD 59457  75 Ub       PM   
677       4855 Peter Bridge\nJohnsonberg, PA 90599-0009  62 Nx       AM   
766             386 Alisha Unions\nSteelebury, ND 19782  28 pJ       PM   
1225    916 Amanda Heights\nNew Johnland, CA 52112-8572  09 vg       PM   
1381  79284 Lisa Mews Suite 069\nKellyborough, CT 71...  45 MR       PM   
1385  67796 James Keys Suite 656\nSouth Katieshire, ...  06 ia       PM   
1568       76108 Barker Manors\nEast Amy, PW 49544-5921  82 Qw       AM   
1727  478 Anita Hill Apt. 766\nAverymouth, FM 50629-...  94 qM       AM   
2146           557 Barton Harbor\nFuentesport, VI 29961  96 MV       AM   
2264   040 Ingram Way Suite 602\nEast Matthew, AR 67341  52 WA       AM   
2267  216 Olivia Court Apt. 439\nTonyshire, IN 19159...  11 za       PM   
2642  899 Williams Prairie Suite 234\nEast Hunterbur...  46 hb       PM   
2828  20135 Miller Green Apt. 763\nGarciaville, KY 6...  79 El       PM   
3291             58939 Reese Cove\nSouth Ross, MA 54531  50 cD       AM   
3349  1984 Christopher Turnpike Suite 904\nRichardbu...  30 xs       AM   
3619  03233 Angela Road Apt. 808\nEast Jilliantown, ...  41 nH       PM   
4071  8161 Craig Passage Apt. 939\nPort Emilystad, R...  45 ab       AM   
4569  88284 Austin Summit Apt. 805\nKellerside, WY 6...  03 WL       AM   
5385                           USNS Smith\nFPO AA 37829  80 BX       AM   
5655       38674 Mckay Vista\nWilsonbury, MO 55611-8663  95 aB       PM   
5692                   PSC 8264, Box 8124\nAPO AE 19700  97 sG       PM   
5814              Unit 5232 Box 4650\nDPO AE 28335-7543  71 So       PM   
5875      8256 Anderson Forest\nFordbury, PA 81027-8980  17 Oe       AM   
5895  87003 Jason Tunnel Apt. 216\nDariusville, MS 8...  72 gL       AM   
6026                   Unit 4385 Box 1428\nDPO AE 04110  58 Ex       AM   
6524  344 Rivera Shore Apt. 224\nJoshuaborough, ND 3...  44 SM       AM   
7239        81561 Stein Rue\nEast Melissaport, MA 64959  45 sZ       PM   
7381  40804 Belinda Roads Apt. 186\nWest Nicoleberg,...  30 GH       AM   
7416    1082 Jason Alley Suite 707\nWhitestad, SC 97805  08 bm       PM   
7514    678 David Manor Suite 323\nBobbymouth, PW 59942  91 NA       AM   
8233  327 Hamilton Hollow\nEast Nicolechester, RI 65868  53 dE       PM   
8381  274 Wallace Pine Apt. 189\nMartinberg, WY 7548...  32 rI       PM   
8408                   Unit 4967 Box 4574\nDPO AP 11412  05 zs       PM   
8434  474 Gonzalez Meadow Suite 581\nEast Vincent, M...  14 St       PM   
8817          3658 Campos Lodge\nChambersfurt, MH 86260  24 ej       PM   
9168  18782 Hall Row Suite 810\nSouth Wesleyville, I...  15 QU       PM   
9856  058 Miranda Locks Suite 792\nPort Davidstad, P...  73 IW       PM   

                                           Browser Info  \
9     Mozilla/5.0 (X11; Linux i686; rv:1.9.7.20) Gec...   
280   Mozilla/5.0 (Windows NT 5.01) AppleWebKit/5362...   
372   Opera/8.42.(X11; Linux x86_64; en-US) Presto/2...   
677   Opera/9.49.(Windows 98; Win 9x 4.90; en-US) Pr...   
766   Opera/8.47.(X11; Linux x86_64; sl-SI) Presto/2...   
1225  Mozilla/5.0 (X11; Linux i686; rv:1.9.7.20) Gec...   
1381  Mozilla/5.0 (Macintosh; PPC Mac OS X 10_5_5; r...   
1385  Mozilla/5.0 (Windows NT 5.1) AppleWebKit/5330 ...   
1568  Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/53...   
1727  Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/53...   
2146  Mozilla/5.0 (compatible; MSIE 6.0; Windows 98;...   
2264  Mozilla/5.0 (Windows NT 6.0) AppleWebKit/5341 ...   
2267  Mozilla/5.0 (Macintosh; PPC Mac OS X 10_6_4; r...   
2642  Mozilla/5.0 (compatible; MSIE 7.0; Windows NT ...   
2828  Mozilla/5.0 (Windows; U; Windows CE) AppleWebK...   
3291  Mozilla/5.0 (compatible; MSIE 9.0; Windows 95;...   
3349  Mozilla/5.0 (X11; Linux x86_64; rv:1.9

- Now I just need the counts 

In [141]:
ecom[(ecom['CC Provider']=='American Express') & (ecom['Purchase Price']>95)].count()

Address             39
Lot                 39
AM or PM            39
Browser Info        39
Company             39
Credit Card         39
CC Exp Date         39
CC Security Code    39
CC Provider         39
Email               39
Job                 39
IP Address          39
Language            39
Purchase Price      39
dtype: int64

## Alternative method

In [143]:
len(ecom[(ecom['CC Provider']=='American Express') & (ecom['Purchase Price']>95)].index)

39

- How many people have a credit card that will expire in 2025? 

In [148]:
ecom['CC Exp Date']

0       02/20
1       11/18
2       08/19
3       02/24
4       10/25
        ...  
9995    03/22
9996    07/25
9997    05/21
9998    11/17
9999    02/19
Name: CC Exp Date, Length: 10000, dtype: object

- Lets check if it's a string data type

In [150]:
ecom['CC Exp Date'].iloc[0]

'02/20'

- Now to grab a string that is the year for example 2020

In [152]:
ecom['CC Exp Date'].iloc[0][3:]

'20'

In [154]:
ecom['CC Exp Date'].apply(lambda exp: exp[3:]=='25')

0       False
1       False
2       False
3       False
4        True
        ...  
9995    False
9996     True
9997    False
9998    False
9999    False
Name: CC Exp Date, Length: 10000, dtype: bool

- Check how many of these are True 

In [156]:
sum(ecom['CC Exp Date'].apply(lambda exp: exp[3:]=='25'))

1033

## Alternative Method

In [158]:
ecom[ecom['CC Exp Date'].apply(lambda exp: exp[3:]=='25')].count()

Address             1033
Lot                 1033
AM or PM            1033
Browser Info        1033
Company             1033
Credit Card         1033
CC Exp Date         1033
CC Security Code    1033
CC Provider         1033
Email               1033
Job                 1033
IP Address          1033
Language            1033
Purchase Price      1033
dtype: int64

- What are the top 5 most popular email providers/hosts (e.g. gmail.com, yahoo.com, etc...)

In [160]:
ecom['Email']

0                    pdunlap@yahoo.com
1                   anthony41@reed.com
2       amymiller@morales-harrison.com
3          brent16@olson-robinson.info
4          christopherwright@gmail.com
                     ...              
9995            iscott@wade-garner.com
9996                mary85@hotmail.com
9997                 tyler16@gmail.com
9998           elizabethmoore@reid.net
9999             rachelford@vaughn.com
Name: Email, Length: 10000, dtype: object

- Notice here that we basically have strings as the data type

In [162]:
example_email = ecom['Email'].iloc[0]

In [163]:
example_email

'pdunlap@yahoo.com'

- Create a list with split() function 

In [165]:
example_email.split('@')

['pdunlap', 'yahoo.com']

- Grab the domain name of the email provider

In [167]:
example_email.split('@')[1]

'yahoo.com'

- Now I can use this logic in a 'lambda expression to grab the email providers'

In [169]:
ecom['Email'].apply(lambda email: email.split('@')[1])

0                  yahoo.com
1                   reed.com
2       morales-harrison.com
3        olson-robinson.info
4                  gmail.com
                ...         
9995         wade-garner.com
9996             hotmail.com
9997               gmail.com
9998                reid.net
9999              vaughn.com
Name: Email, Length: 10000, dtype: object

- Now get the number of counts 

In [171]:
ecom['Email'].apply(lambda email: email.split('@')[1]).value_counts()

hotmail.com             1638
yahoo.com               1616
gmail.com               1605
smith.com                 42
williams.com              37
                        ... 
roberts-petersen.biz       1
jackson-smith.com          1
leblanc-price.info         1
welch.biz                  1
henderson-graves.com       1
Name: Email, Length: 3416, dtype: int64

- Top 5 most popular 'email providers are:

In [174]:
ecom['Email'].apply(lambda email: email.split('@')[1]).value_counts().head(5)

hotmail.com     1638
yahoo.com       1616
gmail.com       1605
smith.com         42
williams.com      37
Name: Email, dtype: int64

## Great Job!